src/cargo/ops/cargo_package.rs

```diff
@@ -178,34 +250,100 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
             to_package: ops::Packages::Default,
-            cli_features,
             ..opts.clone()
         };
-        let ar_files = build_ar_list(ws, pkg, &opts)?;
+        let ar_files = build_ar_list(ws, &pkg, &opts)?;
 
         if opts.list {
-            for ar_file in ar_files {
+            for ar_file in &ar_files {
                 drop_println!(ws.gctx(), "{}", ar_file.rel_str);
             }
         } else {
-            let tarball = create_package(ws, pkg, ar_files)?;
-            if opts.verify {
-                run_verify(ws, pkg, &tarball, &opts)
-                    .with_context(|| "failed to verify package tarball")?;
-            }
-            dsts.push(tarball);
+            let tarball = create_package(ws, &pkg, ar_files, Some(&local_reg))?;
+            local_reg.add_package(&pkg, &tarball)?;
+            outputs.push((pkg, opts, tarball));
         }
     }
 
-    if opts.list {
-        // We're just listing, so there's no file output
-        Ok(None)
-    } else {
-        Ok(Some(dsts))
+    // Verify all packages in the workspace. This can be done in any order, since the dependencies
+    // are already all in the local registry overlay.
+    if opts.verify {
+        for (pkg, opts, tarball) in &outputs {
+            run_verify(ws, pkg, tarball, Some(&local_reg), opts)
+                .with_context(|| "failed to verify package tarballs")?
+        }
+    }
+
+    Ok(outputs.into_iter().map(|x| x.2).collect())
+}
+
+/// Just the part of the dependency graph that's between the packages we're packaging.
+/// (Is the package name a good key? Does it uniquely identify packages?)
+#[derive(Clone, Debug, Default)]
+struct LocalDependencies {
+    packages: HashMap<PackageId, (Package, CliFeatures)>,
+    graph: Graph<PackageId, ()>,
+}
+
+impl LocalDependencies {
+    fn sort(&self) -> Vec<(Package, CliFeatures)> {
+        self.graph
+            .sort()
+            .into_iter()
+            .map(|name| self.packages[&name].clone())
+            .collect()
+    }
+}
+
+/// Build just the part of the dependency graph that's between the given packages.
+fn local_deps(
+    ws: &Workspace<'_>,
+    packages: impl Iterator<Item = (Package, CliFeatures)>,
+) -> LocalDependencies {
+    let packages: HashMap<PackageId, (Package, CliFeatures)> =
+        packages.map(|pkg| (pkg.0.package_id(), pkg)).collect();
+
+    let mut graph = Graph::new();
+    for (pkg, _features) in packages.values() {
+        graph.add(pkg.package_id());
+        for dep in pkg.dependencies() {
+            let dep_id = dep.source_id();
+            if dep.kind() == DepKind::Development || !dep_id.is_path() {
+                continue;
+            };
+
+            let Ok(dep_path) = dep_id.url().to_file_path() else {
+                continue;
+            };
+            let dep_manifest = dep_path.join("Cargo.toml");
+            let Ok(dep_pkg) = ws.load(&dep_manifest) else {
+                continue;
+            };
```

Comment:
I feel like there should be a better way of doin this as appending `Cargo.toml` will cause us problems when we have cargo script.

Can we instead walk workspace members to see if we can find one with the same source?  I feel like that will also be less likely to cause problems in other ways

Reply:
Yes, that works!